packages/lib-classifier/src/plugins/drawingTools/experimental/models/marks/FreehandLine/FreehandLine.js

```diff
@@ -3,277 +3,646 @@ import { FreehandLine as FreehandLineComponent } from '@plugins/drawingTools/com
     },
 
-    shortenPath() {
-      let lengthToRemove = 20
-      while (lengthToRemove--) {
-        self.pathX.pop()
-        self.pathY.pop()
+    inactive() {
+      while (['splice-drag-point', 'splice-close-point', 'splice-append'].indexOf(self.drawActions.at(-1).type) > -1) {
+        self.undo()
+        self.redoActionsClear()
       }
     },
 
-    splice(startIndex, endIndex) {
-      let firstPoint = Math.min(startIndex, endIndex)
-      let lastPoint = Math.max(startIndex, endIndex)
-      /*
-        Reverse the path, if necessary, so that the first point clicked
-        is always the draggable point.
-      */
-      if (firstPoint === endIndex) {
-        const lastIndex = self.points.length - 1
-        self.pathX.reverse()
-        self.pathY.reverse()
-        firstPoint = lastIndex - startIndex
-        lastPoint = lastIndex - endIndex
+    finish() {
+      if (self.points.length < self.minimumPoints) {
+        return self.tool.deleteMark(self)
+      }
+
+      if (self.isDragging) {
+        self.appendPathEnd()
+        self.finished = false
+      } else {
+        self.pathX = []
+        self.pathY = []
+        self.points.forEach(point => {
+          self.pathX.push(point.x)
+          self.pathY.push(point.y)
+        })
       }
-      const deleteCount = lastPoint - firstPoint - 1
-      const clippedPoints = self.points.slice(firstPoint, lastPoint + 1)
-      self.clipPath = clippedPoints.map(({ x, y }) => ({ x, y }))
-      self.pathX.splice(firstPoint + 1, deleteCount)
-      self.pathY.splice(firstPoint + 1, deleteCount)
-      // Make the ends of the spliced section draggable.
-      self.dragPoint = self.points[firstPoint]
-      self.targetPoint = self.points[firstPoint + 1]
-    },
-
-    trim(startIndex, endIndex) {
-      let firstPoint = Math.min(startIndex, endIndex)
-      let lastPoint = Math.max(startIndex, endIndex)
-      /*
-        Reverse the path, if necessary, so that the first point clicked
-        is always at the end of the new line. The end of an open line
-        is the draggable point.
-      */
-      if (lastPoint === endIndex) {
-        const lastIndex = self.points.length - 1
-        self.pathX.reverse()
-        self.pathY.reverse()
-        lastPoint = lastIndex - startIndex
-        firstPoint = lastIndex - endIndex
+    },
+
+    toData() {
+      return {
+        scale: self.scale,
+        points: self.points,
+        dragPoint: self.dragPoint,
+        closePoint: self.closePoint,
+        pathIsClosed: self.pathIsClosed,
+        isDragging: self.isDragging,
+
+        spliceActive: self.spliceActive,
+        spliceReverse: self.spliceReverse,
+        spliceThroughBeginning: self.spliceThroughBeginning,
+        spliceDragPointIndex: self.spliceDragPointIndex,
+        spliceClosePointIndex: self.spliceClosePointIndex,
+        splicePoints: self.splicePoints,
+
+        drawActions: self.drawActions,
+        redoActions: self.redoActions,
       }
```

Comment:
This doesn't seem to be used anywhere.

Reply:
It doesn't change the model state either, so isn't an action. Functions that return model data are views.

Reply:
This is used in the testing suite, specifically for debug purposes. I've moved it up to being a view (good call).